core/src/test/scala/unit/kafka/log/LogTest.scala

```diff
@@ -1013,6 +1013,54 @@ class LogTest extends JUnitSuite {
   }
 
   @Test
+  def testOverCompactedLogRecovery(): Unit = {
+    // append some messages to create some segments
+    val logProps = new Properties()
+    logProps.put(LogConfig.SegmentBytesProp, 1000: java.lang.Integer)
+    logProps.put(LogConfig.MaxMessageBytesProp, 64*1024: java.lang.Integer)
+    logProps.put(LogConfig.IndexIntervalBytesProp, 1: java.lang.Integer)
+    val config = LogConfig(logProps)
+    val log = new Log(logDir,
+      config,
+      recoveryPoint = 0L,
+      time.scheduler,
+      time)
+    val set1 = new ByteBufferMessageSet(
+      compressionCodec = NoCompressionCodec,
+      offsetSeq = Seq(0),
+      messages = new Message("v1".getBytes(), "k1".getBytes(), Message.NoTimestamp, magicValue = Message.CurrentMagicValue))
+    val set2 = new ByteBufferMessageSet(
```

Comment:
This is a great example. Now, I am wondering if we have fixed the issue completely. What happens if we append two messages at offset 0 and at offset Integer.MAX_VALUE.toLong + 2 in succession? On appending the second message, we will first roll a new segment at base offset 1, but the appending of the second message will fail since the offset difference is larger than Integer.MAX_VALUE.